### ISSUE TITLE
Render function appears to not fully execute when it is called in an inactive browser tab

### ISSUE THREAD

**Issue by woranl** [mozilla/pdf.js#6460]
The render function appears to pause (i.e. not fulfilling its promise or not rejecting) when it is executed in an inactive (i.e. background) browser tab.  Once the browser tab becomes active again, the render function will then complete its execution and fulfill its promise.

I was trying to create a pdf thumbnail, then upload the pdf file and the newly created thumbnail to a server.  I noticed that if I select the pdf file for upload and immediately switch to another browser tab, the inactive browser tab will continue to execute its javascript up to the pdf render function.  The render promise will not get fulfilled (i.e. as if javascript pauses), until the browser tab becomes active again.  

In other words, the pdf upload cannot complete in the background because the thumbnail creation pauses and pending for the render promise to get fulfilled.

**Comment by Rob--W** [comment]
`requestAnimationFrame` is used to schedule renderings: https://github.com/mozilla/pdf.js/blob/f06aa6a3802ef6fb6c8919ef3de49bcfe774cb51/src/display/api.js#L1639-L1642

This callback is not invoked when the tab is inactive.

**Comment by woranl** [comment]
Thanks @Rob--W .  I can workaround the problem by strictly using `Promise.resolve(undefined).then(this._nextBound);` to schedule rendering instead of using `window.requestAnimationFrame(this._nextBound);`.  I am not familiar with `requestAnimationFrame`.  What is the implication of not using it to schedule rendering?

**Comment by Rob--W** [comment]
@woranl `Promise.resolve().then` runs the callback as soon as the stack is empty, whereas requestAnimationFrame executes before a repaint. The latter occurs with a lower frequency. By using the former, your page will greedily use CPU until all PDF rendering tasks have completed, even if the user has switched to a different tab. And also, scrolling might not be as smooth compared to rAF.

**Comment by Rob--W** [comment]
There is no bug here, so I'll close this ticket.

**Comment by woranl** [comment]
@Rob--W , what you consider a figure is a bug in my use case. Can anything be done to make pdf.js more versatile?

**Comment by FyrikCN** [comment]
Hi there,

I'm facing the same issue now as it seems to be not loading while I switch to another tab, though it does when I switch back. My PDF-JS version is 2.10.377 so the structure might has changed a bit. I tried to get the page keep rendering when the tab is inactive by using the following code
`
var _this17 = this;

if (this._useRequestAnimationFrame) {
          window.requestAnimationFrame(function () {
            **Promise.resolve(undefined).then(_this17._nextBound);**
          });
        } else {
          Promise.resolve().then(this._nextBound)["catch"](this.cancel.bind(this));
        }
`
to replace the origin code in line 2408-2416 in the file node_modules/pdfjs-dist/lib/display/api.js but it didn't work for me. @Rob--W , am I doing the right thing and why wouldn't it work?

**Comment by corwin-of-amber** [comment]
I have a similar problem when the browser window is occluded by another window. Chrome (correctly) does not fire animation frame events in this case. I am using pdf.js in an NWjs app that also acts as a server, so it's bothersome that the server window must be in the foreground in order for rendering to take place.

**Comment by corwin-of-amber** [comment]
This works for me! amaze.

**Comment by FyrikCN** [comment]
@capnmidnight Thanks! It works perfectly for me :)

**Comment by pdfitdave** [comment]
@capnmidnight - thank you so, so much for the answer. If I encountered this issue before 2022, I would have never been able to fix this issue.
